### DISCUSSION TITLE
How to render large amount of complex polygonal shapes efficiently

### DISCUSSION THREAD

**Discussion by RecursiveVoid** [pixijs/pixijs#10249] [Q&A] [answered]
Hi, my first question here heh.. quite exited... Anyways, i will hop on to my issue here: 

I want to render around 100k of complex polygonal shapes (closed shapes with filling and stroke) , arcs etc.. 

I already experimented and amazed by the power of the pre-rendered sprites. I was using Sprites for rectangle shapes, referencing only one 1x1 texture which i created. Then pushing them to the ParticleContainer. Amazing performance. 

 Unfortunately when it comes to complex polygonal shapes it gets more complicated. Using Graphics is very costly, same with rendering Mesh (compared to sprites) (unless you cache them as bitmaps). 
 
Caching as bitmap is an issue in this case because i want to use viewport and zoom in and out and preserve the scaling of the objects. Which, it works pretty good with basic rectangles and lines in this case. (I recalculate only the height and width and update the shape) 

Any suggestions & tips?

**Comment by bigtimebuddy** [comment, marked as answer]
Rendering Fast Graphics with PixiJS https://medium.com/@bigtimebuddy/rendering-fast-graphics-with-pixijs-6f547895c08c

Maybe this article I wrote can help you.

**Reply by RecursiveVoid** [reply]
Hey! Thank you so much for the answer and the amazing article, there it is used RenderTexture. Its very clever solution with multiple drawn objects with same shape. 

In this case, the challenge is that all Polygonal shapes will have different geometrical properties. 

I was thinking what if i create  a texture from the polygon shape using Graphics, and then the problem will be zooming in and out, i need to recalculate once again and generate texture.. which is [put skull emoji here]

**Reply by RecursiveVoid** [reply]
Your article kinda inspired me; what if i create the polygonal shape using graphics, (scaled up version) and create a texture from it.. so in zoom in it won't lose any quality.  (but will cost a lot of memory later on with more different complex shapes) 

I'm still skeptical about using graphics or mesh in this case. Actually with mesh i can create the vertices in the shader.. but i really don't know how it will impact the GPU with **that** amount of drawn objects.

**Reply by bigtimebuddy** [reply]
100K shapes is a lot. If you don't need all those on screen at once you should definitely try to avoid it. The more you push the performance, the more sensitive things become to very minor changes. Even if you are using a single sprite/render texture, you will start to bump into issues with fillrate (the amount of pixels that need be painted per-frame). Or if you're using a lot of unique shapes that could easily create memory issues or break Sprite batching (most devices have a max texture limit size ~16).

**Reply by RecursiveVoid** [reply]
Exactly like in your article, I should consider about culling. And also maybe an intelligent layering system in container where it stops rendering the overlapping Shaders and textures which are not visible from first place. 

Alright, i will mark this as the Answer since it was super helpful. 
Thank you Matt.

**Reply by Amansaxena001** [reply]
@bigtimebuddy Even If we enable culling ( which I ve enabled on my app ) which does give some boost but when we have a scenario of many unique shapes ( more importantly the shapes can be changed on the fly ) that are being drawn using `Pixi.Graphics` and if you add borders to that then its a mess as the FPS drops are huge.

Any suggestions from your side on what can be done here?

**Reply by bigtimebuddy** [reply]
Consider rendering Graphics to a RenderTexture. Adding stokes can dramatically increase the geometry making it more sensitive to performance with a large number of shapes. Using a RenderTexture basically treats it like a Sprite (rectangle with 4 vertices). Alternative option, that you should read the article above if you haven't already, is to draw the Graphics at a smaller size and then scale up. This will reduce the number of point for triangulating. Good luck.

**Reply by Amansaxena001** [reply]
I am having a hard time using `RenderTexture` to render dynamic graphics. When I use renderTexture and use the sprite to dynamically set the x, y, width, height of the container the sprite is scaled too much especially when setting `width` &  `height`. For eg the calculated layout values are 
```
  const pixiComp: PIXI.Sprite = new PIXI.Sprite(commonTexture); -> commonTexture is a `RenderTexture`( ref your blog )
  
  const { x, y, width, height } = calculatedLayoutValues

// these are getting set correctly
    pixiComp.x = x;
    pixiComp.y = y;
    pixiComp.tint = some color

// when I set these the whole sprite is scaled too much instead the expectation was to look like exactly how the graphics would look like if pass same props.
    pixiComp.width = width
    pixiComp.height = height;

```

Note- This exact implementation works fine when using graphics but when using sprites instead of graphics its not working as expected.

@bigtimebuddy

**Reply by bigtimebuddy** [reply]
Seems like you need more help @Amansaxena001 with your project than we are able to support on these forums. We unfortunately cannot help all users debug their specific code. I would suggest you join the Discord (link on readme), you're more likely to get helpful community members there. Also, please make sure to read the guides on PixiJS.com as well as https://www.pixijselementals.com/#introduction. When in doubt try ChatGPT, StackOverflow and Google.
